### PR TITLE
Registry vocabulary

### PR DESCRIPTION
**Pull request by cguardia**
@davisagli @alecpm this is the registry vocabulary work. Please take a look.

Thanks.

### CODE REVIEW

File: pleiades/vocabularies/browser/controlpanel.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    description = u""
+    enableCSRFProtection = True
+
+    def updateFields(self):
+        super(PleiadesSettingsEditForm, self).updateFields()
+        self.fields['time_periods'].widgetFactory = DataGridFieldFactory
+        self.fields['time_periods'].allow_insert = True
+
+    def updateWidgets(self):
+        super(PleiadesSettingsEditForm, self).updateWidgets()
```
Comment (davisagli) [inline pleiades/vocabularies/browser/controlpanel.py]: This looks unnecessary.

File: pleiades/vocabularies/configure.zcml [hunk 2]
```diff
@@ -1,11 +1,14 @@
...
   xmlns="http://namespaces.zope.org/zope"
   xmlns:five="http://namespaces.zope.org/five"
+  xmlns:grok="http://namespaces.zope.org/grok"
   xmlns:genericsetup="http://namespaces.zope.org/genericsetup"
   i18n_domain="pleiades.vocabularies"
   >
 
   <five:registerPackage package="." initialize=".initialize" />
+  <includeDependencies package="." />
+  <grok:grok package="." />
```
Comment (davisagli) [inline pleiades/vocabularies/configure.zcml]: Pleiades doesn't use grok and I don't think we should add it, given that it's not widely used in the Plone community. We can register the vocab using a ZCML `utility` directive instead.

**Comment by davisagli** [comment]
I guess it wasn't clear from the stories but there are some other things that need to happen so that the new vocabulary is useful:
1. We need to find all the places in the code that use the old time-periods vocabulary and update them to use the new vocabulary.
2. We need to migrate the existing time period values to the new vocabulary.

The vocabulary terms need an id that is separate from the title and description. For example here is one of the current vocab items: http://pleiades.jazkarta.com/vocabularies/time-periods/early-geometric ... it has an id of 'early-geometric' (which is the value actually stored on content items), a title of "Early Geometric (Greek; 900-850 BC)" and a description of "The Early Geometric period in ancient Greece, from 900-850 before the birth of Christ. [[-900, -850]]". (The values in square brackets are the lower/upper bound and can be removed from the description as part of the migration.)

**Comment by davisagli** [comment]
Also, I'm not sure that excluding hidden terms from the vocabulary is the right approach. If someone is editing a content item that uses one of the hidden terms, we still want to show its title, which I think means it needs to be in the vocab. Instead, I think we probably need to customize the Archetypes widget used for the time period field so that it only shows vocabulary terms that are a) non-hidden or b) the current value.

**Comment by davisagli** [comment]
This is a good start though.

File: pleiades/vocabularies/browser/controlpanel.py (hunk 1 above)
Comment (cguardia) [inline pleiades/vocabularies/browser/controlpanel.py]: So, they are not going to edit the vocabularies from the Plone site at all?

Comment (davisagli) [inline pleiades/vocabularies/browser/controlpanel.py]: They will. I meant that the `updateWidgets` function that only calls `super` is unnecessary.

**Comment by cguardia** [comment]
@davisagli the terms do have an id that is separate from the description.

**Comment by cguardia** [comment]
@davisagli or do you mean, not use SimpleTerm for the terms to allow passing desc and hidden values?

**Comment by davisagli** [comment]
@cguardia I just overlooked the id, sorry. For the vocabulary terms, we should use our id as the term's `value` and `token`, and our title (not description) as the term's `title`

File: pleiades/vocabularies/vocabularies.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    def vocabulary_factory(context):
+        registry = getUtility(IRegistry)
+        values = registry.get(path, [])
+        values.sort()
+        terms = []
+        for value in values:
+            terms.append(SimpleTerm(
+                value=value['id'],
+                token=value['id'].encode('raw_unicode_escape'),
```
Comment (alecpm) [inline pleiades/vocabularies/vocabularies.py]: Is this encoding appropriate? I thought token needed to be ASCII, maybe use:

```
from binascii import b2a_qp
...
b2a_qp(value['id'].encode('utf-8'))
```

Comment (davisagli) [inline pleiades/vocabularies/vocabularies.py]: raw_unicode_escape should produce 7-bit ascii, with other characters encoded in python literal notation

Comment (alecpm) [inline pleiades/vocabularies/vocabularies.py]: Hm, I think produces 8-bit output equivalent to latin-1 with escapes for other characters.

```
>>> u'\u00e9'.encode('raw_unicode_escape')
'\xe9'

>>> print u'\u00e9'.encode('raw_unicode_escape')
?
```

Seems like `unicode_escape` does escaped literals though:

```
>>> u'\u00e9'.encode('unicode_escape')
'\\xe9'
>>> print u'\u00e9'.encode('unicode_escape')
\xe9

>>> b2a_qp(u'\u00e9'.encode('utf-8'))
'=C3=A9'
```

Comment (davisagli) [inline pleiades/vocabularies/vocabularies.py]: Oops, thanks!

**Pull request merged**